fix(node-package-tool): default to no (dev) dependencies if unparsable

sternenseemann opened the pull request:
In the node world you can't depend on anything, especially not that a certain field of `package.json` is well-formed.

Since the `dependencies` field is sometimes malformed we default to `{}`. I think packages with malformed `dependencies` field can't reasonably expect their dependencies to be installed correctly.

A negative side effect of this change could be that it'd be hard to debug such an error, since `node-package-tool` will only output a warning getting drowned out be other output of `nix`. This *should* not be an issue however since people won't necessarily use `yarn2nix` for development, so if `yarn` itself behaves similarly the issue would be spotted earlier hopefully.

To demonstrate the original issue apply this [patch](https://gist.github.com/sternenseemann/bdb4b7c5d066f70f670ac09211213320) in [CodiMD/server](https://github.com/codimd/server) and try to build it with `master` `yarn2nix`. Then the issue is triggered by <https://github.com/garycourt/JSV/blob/master/package.json#L29> being a empty array instead of a empty dictionary.

sternenseemann committed:
fix(nodejs/Package.hs): default to no (dev) dependencies if unparsable

In the node world you can't depend on anything, especially not that a
certain field of package.json is well-formed.

The dependencies field is sometimes malformed, probably in most cases it
is safe to default to {}. I have observed this in the wild with JSV 0.4.2
(https://github.com/garycourt/JSV) where dependencies is [] instead of
{} or being missing. I think packages with malformed dependency field
can't reasonably expect their dependencies to be installed correctly.

We only default to {} in cases where we can expect beyond a reasonable
doubt that there are no dependencies being expressed by the field:

* Empty array
* Scalars (Number, String, …)

We fail parsing if the field is malformed and seems to contain
something:

* Non-empty array
* Object that can't be parsed to Dependencies (i. e. is malformed)

We use this strategy for both dependencies and devDependencies.

Profpatsch commented:
I’m thinking, maybe we can make the default just apply to an empty list, not a non-empty list. Because that would really be a case which we don’t know how to handle (and don’t really expect).

So 
* dict -> parse
* empty list or scalar -> warning with no deps
* non-empty list -> crash

sternenseemann commented:
A string should also be a crash, I guess.

Profpatsch commented:
I think strings can be warnings, because there is not much they could mean. A non-empty list could mean the author did something strange.

sternenseemann commented:
I guess a string could be a single dependency or something? unlikely though

sternenseemann commented:
Done in a59324a23c5e02afe5e437d3c1461ca6b539d700.

sternenseemann commented on src/Distribution/Nodejs/Package.hs:
```diff
@@ -98,11 +99,33 @@ instance A.FromJSON LoggingPackage where
+                -- everything else defaults to mempty and generates a warning
+                _ -> putWarning def
+                       (WrongType { wrongTypeField   = field
+                                  , wrongTypeDefault = Just (show def) }))
```
Maybe here it is nicer to use `Just "{}"`, so the output is `defaulting to {}` instead of `fromList []`?

Profpatsch committed:
refactor(tests): move helpers out of test function

Profpatsch committed:
refactor(tests): defunctionalize hasWarning predicate

Profpatsch committed:
feat(tests): add some unit tests for package dependencies

Profpatsch committed:
refactor(nodejs/Package.hs): simplify parseDependecies

The logic is slightly easier to follow if we re-use the `tryWarn`
helper to parse the field as `Aeson.Value`, and then forward the value
to the `parseDependencies` function, which matches on the Value,
and creates warnings and failures for the wrong cases.

Profpatsch commented:
I pushed a small refactor and added some unit tests.

Profpatsch commented:
LGTM from my side, but maybe you want to look over my commits.
